Cython/Compiler/FusedNode.py

```diff
@@ -686,31 +715,61 @@ def __pyx_fused_cpdef(signatures, args, kwargs, defaults):
             self._buffer_declarations(pyx_code, decl_code, all_buffer_types, pythran_types)
             env.use_utility_code(Code.UtilityCode.load_cached("Import", "ImportExport.c"))
             env.use_utility_code(Code.UtilityCode.load_cached("ImportNumPyArray", "ImportExport.c"))
+        
+        context.update(
+            global_sigindex_name = fused_cpdef_globalindex,
+            func_sigindex_key = env.mangle(fused_cpdef_globalindex_prefix, orig_py_func.entry.cname)
+        )
 
         pyx_code.put_chunk(
             u"""
-                candidates = []
-                for sig in <dict>signatures:
-                    match_found = False
-                    src_sig = sig.strip('()').split('|')
-                    for i in range(len(dest_sig)):
-                        dst_type = dest_sig[i]
-                        if dst_type is not None:
-                            if src_sig[i] == dst_type:
-                                match_found = True
-                            else:
-                                match_found = False
-                                break
+                global {{global_sigindex_name}}
+                if '{{global_sigindex_name}}' not in <dict>globals():
+                    {{global_sigindex_name}} = {}
+                if '{{func_sigindex_key}}' not in <dict>{{global_sigindex_name}}:
+                    {{global_sigindex_name}}['{{func_sigindex_key}}'] = {}
+                _fused_sigindex = (<dict>{{global_sigindex_name}})['{{func_sigindex_key}}']
+            """
+        )
 
-                    if match_found:
-                        candidates.append(sig)
+        self._fused_signature_index(pyx_code)
 
+        pyx_code.put_chunk(
+            u"""
+                sigindex_matches = []
+                sigindex_candidates = [_fused_sigindex]
+                
+                for dst_type in <list>dest_sig:
+                    found_matches = []
+                    found_candidates = []
+                    # Make two seperate lists: One for for signature sub-trees
+                    #        with at least one definite match, and another for
+                    #        signature sub-trees with only ambiguous matches
+                    #        (where `dest_sig[i] is None`).
+                    if dst_type is None:
+                        for sn in <list>sigindex_matches:
```

Comment:
Would it make sense to make this `cdef list sigindex_matches` at the start of the function? Then you wouldn't need the casts at every use. (Similarly for other variables).

The downside is that this might generate more type checks when the variables are assigned, which could be why it wasn't done in the first place.

Reply:
Good call.

I think the assignment type checks should probably be fine, since they're basically all to either literals or to other variables that can be statically typed.

Reply:
I think I originally used type casts for 1) otherwise untyped arguments that were passed into the function (there's a bug that prevented typing them) and 2) loop variables for values of a known type. Since this is generated code, I think type cases are just fine wherever helpful. They can sometimes lead to less overhead because they avoid type checks that may otherwise occur on assignments to typed variables.